numpy/core/src/multiarray/ctors.c

```diff
@@ -782,108 +781,32 @@ discover_dimensions(PyObject *obj, int *maxndim, npy_intp *d, int check_it,
         return 0;
     }
 
-    /* obj is a PEP 3118 buffer */
-    /* PEP 3118 buffer interface */
-    if (PyObject_CheckBuffer(obj) == 1) {
```

Comment:
Does the removal of this indicate that it gets checked anyway as part of the `_array_from_array_like` code path?

If so, does this fix the dtype inference from a list of ctypes? I'll try and find the issue...

Reply:
#13628

Reply:
Thanks, I have to run some errands, but will look at it. That might be a reason for just doing the simple thing and using it in the dtype discovery as well.

Reply:
Hmm, no. If I plug that in into the dtype discovery, all that changes is that the warning is given more times...

Reply:
The warning is expected - but you still get the `ValueError`?

Reply:
Yeah

Reply:
> Does the removal of this indicate that it gets checked anyway as part of the `_array_from_array_like` code path?

Just checked, this does take the other code path. It's a little slower because it constructs the full array for PEP3118 types, but its still O(1), and if we can get caching working then that won't matter.

Reply:
I have similar code paths with caching in my dtypemeta branch, but it is a bit different (and needs cleanups right now). But we can definitely do it. One note about caching: It does mean holding possibly a bit more things in memory.